sbt_common/messaging/src/main/scala/uk/ac/wellcome/messaging/message/MessageStream.scala

```diff
@@ -26,60 +19,19 @@ class MessageStream[T] @Inject()(actorSystem: ActorSystem,
                                  s3Client: AmazonS3,
                                  messageReaderConfig: MessageReaderConfig,
                                  metricsSender: MetricsSender)
-    extends Logging {
-  val decider: Supervision.Decider = {
-    case _: Exception => Supervision.Resume
-    case _ => Supervision.Stop
-  }
-  implicit val system = actorSystem
-  implicit val materializer = ActorMaterializer(
-    ActorMaterializerSettings(system).withSupervisionStrategy(decider))
-  implicit val dispatcher = system.dispatcher
+    extends SQSStream[T](
+      actorSystem = actorSystem,
+      sqsClient = sqsClient,
+      sqsConfig = messageReaderConfig.sqsConfig,
+      metricsSender = metricsSender
+    ) {
 
   val s3ObjectStore = new S3ObjectStore[T](
     s3Client = s3Client,
     s3Config = messageReaderConfig.s3Config
   )
 
-  def foreach(streamName: String, process: T => Future[Unit])(
-    implicit decoderT: Decoder[T]): Future[Done] =
-    SqsSource(messageReaderConfig.sqsConfig.queueUrl)(sqsClient)
-      .mapAsyncUnordered(10) { message =>
-        val metricName = s"${streamName}_ProcessMessage"
-        metricsSender.timeAndCount(
-          metricName,
-          () => readAndProcess(streamName, message, process))
-      }
-      .map { m =>
-        (m, MessageAction.Delete)
-      }
-      .runWith(SqsAckSink(messageReaderConfig.sqsConfig.queueUrl)(sqsClient))
-
-  private def readAndProcess(
-    streamName: String,
-    message: Message,
-    process: T => Future[Unit])(implicit decoderT: Decoder[T]) = {
-    val processMessageFuture = for {
-      t <- read(message)
-      _ <- process(t)
-    } yield message
-
-    processMessageFuture.onFailure {
-      case exception: GracefulFailureException =>
-        logger.warn(s"Failure processing message", exception)
-      case exception: Exception =>
-        logger.error(s"Failure while processing message.", exception)
-        metricsSender.incrementCount(
-          s"${streamName}_MessageProcessingFailure",
-          1.0)
-    }
-    processMessageFuture
-  }
-
-  private def read(message: sqs.model.Message)(
-    implicit decoderN: Decoder[NotificationMessage],
-    decoderT: Decoder[T]
-  ): Future[T] = {
+  override def read(message: sqs.model.Message)(implicit decoderT: Decoder[T]): Future[T] = {
```

Comment:
Uhm, one of the things I really don't like in `MessageWorker` and `SqsWorker` is using polymorphism rather than composition. It makes it hard to follow how those functions are used. I'd rather not introduce the same pattern here.
Because there is no mutable state there should be a way of just passing functions around.

Reply:
So instead you have to pass a `read()` method into `foreach`? Is that the idea?

Reply:
To further extend what I mean, I think `MessageStream` should be using `SQSStream` rather than extending it. Its `foreach` function should internally call `SQSStream.foreach` passing it a function that goes from `NotificationMessage` to `Future[Unit]`. That function should be in turn be made of:
- decoding the `MessagePointer` and getting the object from s3
- decoding the object from s3 into T
- apply the `process` function supplied by the user in `MessageStream.foreach`

Reply:
Ah, got it. Yeah, that makes sense.